ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java

```diff
@@ -0,0 +1,294 @@
+ */
+@ApplicationScoped
+public class NarayanaTransactionManager extends DelegatingTransactionManager {
+
+  private final Event<Transaction> transactionScopeInitializedBroadcaster;
+
+  private final Event<Object> transactionScopeDestroyedBroadcaster;
+  
+  /**
+   * Creates a new {@link NarayanaTransactionManager}.
+   *
+   * @param beanManager a {@link BeanManager} to use to find a
+   * relevant {@link TransactionManager} to which to delegate all
+   * operations
+   *
+   * @param transactionScopeInitializedBroadcaster an {@link Event}
+   * for broadcasting the {@linkplain Initialized initialization} of
+   * the {@linkplain TransactionScoped transaction scope}; may be
+   * {@code null}
+   *
+   * @param transactionScopeDestroyedBroadcaster an {@link Event} for
+   * broadcasting the {@linkplain Destroyed destruction} of the {@link
+   * TransactionScoped transaction scope}; may be {@code null}
+   *
+   * @see com.arjuna.ats.jta.TransactionManager#transactionManager()
+   *
+   * @see #begin()
+   *
+   * @see #commit()
+   *
+   * @see #rollback()
+   *
+   * @see TransactionScoped
+   */
+  @Inject
+  public NarayanaTransactionManager(final BeanManager beanManager,
+                                    @Initialized(TransactionScoped.class)
+                                    final Event<Transaction> transactionScopeInitializedBroadcaster,
+                                    @Destroyed(TransactionScoped.class)
+                                    final Event<Object> transactionScopeDestroyedBroadcaster) {
+    super(getDelegate(beanManager));
+    this.transactionScopeInitializedBroadcaster = transactionScopeInitializedBroadcaster;
+    this.transactionScopeDestroyedBroadcaster = transactionScopeDestroyedBroadcaster;
+  }
+
+  private static final TransactionManager getDelegate(final BeanManager beanManager) {
+    Objects.requireNonNull(beanManager);
+    final TransactionManager returnValue;
+    Set<Bean<?>> beans = beanManager.getBeans(TransactionManager.class);
+    if (beans == null || beans.isEmpty()) {
+      // There were no beans registered with TransactionManager as
+      // their bean type, *including this class*.  This is an edge
+      // case, but not as edgy as you might think.  For example, if
+      // someone's portable extension vetoes us but creates us by
+      // hand, we might be in this situation.
+      //
+      // Action: find a TransactionManager in JNDI if possible.
+
+      final InitialContext ic;
+      beans = beanManager.getBeans(InitialContext.class);
+      if (beans == null || beans.isEmpty()) {
+        // There was no bean registered with InitialContext among its
+        // bean types.
+        //
+        // Action: create a new InitialContext by hand with no
+        // arguments.
+        
+        InitialContext temp = null;
+        try {
+          temp = new InitialContext();
+        } catch (final NamingException namingException) {
+          // TODO: log
```

Comment:
If this exception is thrown then the code on line 156 will blow up because ic is set to the value null in the finally block.

Reply:
Nice catch!